service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogHandler.java

```diff
@@ -762,7 +777,190 @@ public LoadTableResponse updateTable(
+
+                  if (found != 1 || !Objects.equals(parentToRollbackTo, parentSnapshotId)) {
+                    // nothing can be done as this implies there was a non replace
+                    // snapshot in between or there is more than setRef ops, we don't know where
+                    // to go.
+                    throw new ValidationFailureException(e);
+                  }
+
+                  // first we should also set back the ref we wanted to set, back to the base
+                  // on which the current update is based on.
+                  metadataBuilder.setBranchSnapshot(parentSnapshotId, ref.name());
+
+                  // apply the remove snapshots update in the current metadata.
+                  // NOTE: we need to setRef to parent first and then apply remove as the remove
+                  // will drop. The tags / branch which don't have reference.
+                  // NOTE: we can skip removing the now orphan base. Its not a hard requirement.
+                  // just something good to do, and not leave for Remove Orphans.
+                  updateToRemoveSnapshot.forEach((update -> update.applyTo(metadataBuilder)));
+                  // Ref rolled back update correctly to snapshot to be committed parent now.
+                  newBase = metadataBuilder.build();
+                  // move the lastSequenceNumber back, to apply snapshot properly.
+                  // Seq number are considered increasing monotonically, snapshot over snapshot, so
+                  // this is important.
+                  Class<?> clazz = newBase.getClass();
+                  try {
+                    Field field = clazz.getDeclaredField("lastSequenceNumber");
+                    field.setAccessible(true);
+                    // this should point to the sequence number that current tip of the
+                    // branch belongs to, as the new commit will be applied on top of this.
+                    field.set(newBase, newBase.currentSnapshot().sequenceNumber());
+                  } catch (NoSuchFieldException | IllegalAccessException ex) {
+                    throw new RuntimeException(ex);
+                  }
+                }
+
+                // double check if the requirements passes now.
+                try {
+                  TableMetadata baseWithRemovedSnaps = newBase;
+                  request
+                      .requirements()
+                      .forEach((requirement) -> requirement.validate(baseWithRemovedSnaps));
+                } catch (CommitFailedException e) {
+                  throw new ValidationFailureException(e);
+                }
+
+                TableMetadata.Builder newMetadataBuilder = TableMetadata.buildFrom(newBase);
+                request.updates().forEach((update) -> update.applyTo(newMetadataBuilder));
+                TableMetadata updated = newMetadataBuilder.build();
+                // always commit this
+                taskOps.commit(base, updated);
```

Comment:
I think this is an interesting approach and probably the right one, but I'm wondering if you considered just reloading the metadata.json from the time of the rollback and applying the commit to that?

Reply:
Posting Discussion offline : Idea was to not skip schema changes that could have potentially landed in between and for which we don't create a new snapshot we just create a new metadata.json so re-using metadata.json was something which we wanted. Though reloading the metadata.json from the time of rollback is also possible we can also checks like if current_schem_id, stats files etc are same or not to ensure we are not loosing any updates which doesn't have snapshot of their own, though this idea looks a bit easy with implementation wise hence choosing above.
TODO: think more on how else can this be achieved.